buildSrc/src/main/groovy/airbyte-docker.gradle

```diff
@@ -0,0 +1,216 @@
+                    return false
+                }
+            }
+
+            if(notUpToDate) {
+                return false;
+            }
+
+            logger.debug "stored " + stored.trim()
+
+            def upToDate = current.equals(stored.trim())
+
+            logger.debug "uptodate " + upToDate.toString()
+
+            return upToDate
+        } else {
+            return false
+        }
+    }
+
+    static def createTask(Project project, String taskName, String dockerFile) {
+        if (project.file(dockerFile).exists()) {
+            def filteredProjectFiles = project.fileTree(project.projectDir).filter {
+                file -> !file.toString().contains(".venv")
+            }
+            project.task(taskName, type: AirbyteDockerTask) {
+                def dockerPath = Paths.get(project.projectDir.absolutePath, dockerFile)
+                def hash = MessageDigest.getInstance("MD5").digest(dockerPath.getBytes()).encodeHex().toString()
+                rootDir = project.rootProject.rootDir
+                projectDir = project.projectDir
+                projectFiles = filteredProjectFiles
+                idFileOutput = project.file(Paths.get(project.rootProject.rootDir.absolutePath, '.dockerversions', hash).toString())
+                baseImageHashes = getBaseImageHashes(project.rootProject.imageToVersion, dockerPath.toFile())
+                dependsOn project.assemble
+
+                outputs.upToDateWhen {
+                    return isUpToDate(logger, idFileOutput, project, dockerPath)
+                }
+            }
+        } else {
+            project.task(taskName) {
+                logger.info "Skipping ${taskName} because ${dockerFile} does not exist."
+            }
+        }
+    }
+
+    void apply(Project project) {
+        // set (and cache) global image to version map
+        project.rootProject.ext.imageToVersion = {
+            if(!project.rootProject.hasProperty("imageToVersion")) {
+                def imageToVersion = [:]
+                def stdout = new ByteArrayOutputStream()
+                project.exec {
+                    commandLine "docker", "images", "--no-trunc", "-f", "dangling=false", "--format", "{{.Repository}}:{{.Tag}} {{.ID}}"
+                    standardOutput = stdout;
+                }
+
+                "$stdout".eachLine { line ->
+                    def splits = line.split()
+                    imageToVersion.put(splits[0], splits[1])
+                }
+
+                return imageToVersion
+            } else {
+                return project.rootProject.imageToVersion
+            }
+        }()
+
+        createTask(project, 'airbyteDockerBuildMain', 'Dockerfile')
+        createTask(project, 'airbyteDockerBuildTest', 'Dockerfile.test')
+
+        project.task('airbyteDocker') {
+            dependsOn project.airbyteDockerBuildMain
+            dependsOn project.airbyteDockerBuildTest
+        }
+
+        project.build.dependsOn project.airbyteDocker
```

Comment:
do we want to build all docker images as part of the build? I guess if this is incremental it's not an issue for speed. Definitely better than remembering to run all the different `buildImage` commands correctly.

Reply:
The way I'm thinking about it is that if the images don't build, then the the build didn't really pass. If `composeBuild` runs as part of the build, why not the base images? It'll be slow if there's something like a base image change. We'll probably need to be more conscious of using exclusions in the future with things like `buildImage`. I'm open to changing this though if you believe this shouldn't be part of the build. 

We could also do something like have a flag for whether or not it's part of the base build, and then we could add all of the bases but not all integrations, which might be the best balance.

Reply:
Actually just going to follow the suggestion https://github.com/airbytehq/airbyte/pull/817#discussion_r519094543 to move integration test image building away only.